cdcpd/src/deformable_object_configuration.cpp

```diff
@@ -21,43 +27,141 @@ void DeformableObjectTracking::makeCloud(Eigen::Matrix3Xf const& points_mat)
+    int i = 0;
+    for (auto const& edge : edges)
+    {
+        edges_new(0, i) = std::get<0>(edge);
+        edges_new(1, i) = std::get<1>(edge);
+        ++i;
+    }
+    edges_ = edges_new;
+}
 
-DeformableObjectTracking RopeConfiguration::makeTemplate(Eigen::Vector3f const& start_position,
-    Eigen::Vector3f const& end_position)
+void DeformableObjectConfiguration::initializeTracking()
 {
-    Eigen::Matrix3Xf template_vertices(3, num_points);  // Y^0 in the paper
-    Eigen::VectorXf thetas = Eigen::VectorXf::LinSpaced(num_points, 0, 1);
-    for (int i = 0; i < num_points; ++i) {
+    DeformableObjectTracking object_tracking = makeTemplate();
+    tracked_ = DeformableObjectTracking(object_tracking);
+    initial_ = DeformableObjectTracking(object_tracking);
+}
+
+RopeConfiguration::RopeConfiguration(int const num_points, float const max_rope_length,
+    Eigen::Vector3f const rope_start_position, Eigen::Vector3f const rope_end_position)
+    : DeformableObjectConfiguration{num_points},
+      max_rope_length_(max_rope_length),
+      rope_start_position_initial_(rope_start_position),
+      rope_end_position_initial_(rope_end_position)
+{
+    max_segment_length_ = max_rope_length / (static_cast<float>(num_points) - 1);
+}
+
+
+DeformableObjectTracking RopeConfiguration::makeTemplate()
+{
+    Eigen::Matrix3Xf template_vertices(3, num_points_);  // Y^0 in the paper
+    Eigen::VectorXf thetas = Eigen::VectorXf::LinSpaced(num_points_, 0, 1);
+    for (int i = 0; i < num_points_; ++i) {
         auto const theta = thetas.row(i);
-        template_vertices.col(i) = (end_position - start_position) * theta + start_position;
+        template_vertices.col(i) = (rope_end_position_initial_ - rope_start_position_initial_)
+            * theta + rope_start_position_initial_;
     }
-    Eigen::Matrix2Xi template_edges(2, num_points - 1);
+    Eigen::Matrix2Xi template_edges(2, num_points_ - 1);
     template_edges(0, 0) = 0;
-    template_edges(1, template_edges.cols() - 1) = num_points - 1;
+    template_edges(1, template_edges.cols() - 1) = num_points_ - 1;
     for (int i = 1; i <= template_edges.cols() - 1; ++i) {
         template_edges(0, i) = i;
         template_edges(1, i - 1) = i;
     }
     DeformableObjectTracking configuration;
-    configuration.vertices = template_vertices;
-    configuration.edges = template_edges;
+    configuration.vertices_ = template_vertices;
+    configuration.edges_ = template_edges;
     configuration.makeCloud(template_vertices);
 
+    return configuration;
+}
+
+ClothConfiguration::ClothConfiguration(float const length_initial, float const width_initial,
+    float const grid_size_initial_guess)
+    : DeformableObjectConfiguration{},
+      length_initial_{length_initial},
+      width_initial_{width_initial},
+      grid_size_initial_guess_{grid_size_initial_guess},
+      num_points_length_{std::round(length_initial / grid_size_initial_guess) + 2},
```

Comment:
+2 is suspicious... maybe should check. if length is 1m and grid size guess is 1m, wouldn't this make num_points 3? why 3 not 2?